README.md
# Traffic Info

Tired of being "suprised" by traffic when you leave your workplace?

This little tool sends you an email containing a map of the area of your choice with traffic informations.

![Traffic info screenshot](https://github.com/Da-Juan/traffic_info/blob/master/docs/_static/images/email_screenshot.png)

Written in Python 3.6 it relies on the following:
* [ChromeDriver](http://chromedriver.chromium.org/)
* [ConfigArgParse](https://github.com/bw2/ConfigArgParse)
* [Jinja2](http://jinja.pocoo.org/) template engine
* [Selenium](https://github.com/SeleniumHQ/Selenium) WebDriver
* [Workalendar](https://pypi.org/project/workalendar/)
* [Google Maps](https://maps.google.com)

This project's source code is available on [Github](https://github.com/Da-Juan/traffic_info).

## User's Guide

### Requirements

ChromeDriver is used to take the screenshot so you need to install [Chromium](https://www.chromium.org/) or [Google Chrome](https://www.google.com/chrome/).
Then, to install ChromeDriver you can use the provided script `bin/install_chromedriver.sh`.

Next you'll need a SMTP server to send the email notifications.

Optionally, you'll need a Google Maps Javascript API key if you have an intensive use of traffic-info.
Refer to [Google Maps Platform documentation](https://developers.google.com/maps/documentation/javascript/get-api-key).

### Install

#### Install from source

Create a Python3 virtualenv then run `setup.py`:

```text
python3 -m venv venv
source venv/bin/activate
python3 setup.py install
```

### Usage

You can run the script from the command line or in a cron to get the traffic infos each day just before leaving your work place.

#### Basic usage

Four parameters are required to run the script:
* the latitude of the center point of the map
* the longitude of the center point of the map
* the email sender’s address
* and the email recipient’s address.

```text
traffic-info -l 43.6037834 -L 1.4402123 -f traffic-info@example.com -t user@example.com
```

#### Command line reference

```text
usage: traffic-info [-h] [-c CONFIG_FILE] [-d WEBDRIVER_PATH] [-k API_KEY] -l LATITUDE -L LONGITUDE [-z ZOOM] -f
                    EMAIL_FROM -t EMAIL_TO [-s SMTP_SERVER] [-p SMTP_PORT] [-S] [-u SMTP_LOGIN] [-w SMTP_PASSWORD]
                    [-W SCREENSHOT_WIDTH] [-H SCREENSHOT_HEIGHT] [-C COUNTRY_CODE]

options:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file path.
  -d WEBDRIVER_PATH, --webdriver_path WEBDRIVER_PATH
                        ChromeDriver's path.
  -k API_KEY, --api_key API_KEY
                        Google Maps Javascript API key.
  -l LATITUDE, --latitude LATITUDE
                        Latitude of the center point of the map.
  -L LONGITUDE, --longitude LONGITUDE
                        Longitude of the center point of the map.
  -z ZOOM, --zoom ZOOM  Google Maps zoom level.
  -f EMAIL_FROM, --email_from EMAIL_FROM
                        Email sender’s address.
  -t EMAIL_TO, --email_to EMAIL_TO
                        Email recipient’s address.
  -s SMTP_SERVER, --smtp_server SMTP_SERVER
                        SMTP server’s address.
  -p SMTP_PORT, --smtp_port SMTP_PORT
                        SMTP server’s port.
  -S, --smtp_use_ssl    Use SMTPS.
  -u SMTP_LOGIN, --smtp_login SMTP_LOGIN
                        SMTP server’s login.
  -w SMTP_PASSWORD, --smtp_password SMTP_PASSWORD
                        SMTP server’s password.
  -W SCREENSHOT_WIDTH, --screenshot_width SCREENSHOT_WIDTH
                        Screenshot’s width.
  -H SCREENSHOT_HEIGHT, --screenshot_height SCREENSHOT_HEIGHT
                        Screenshot’s height.
  -C COUNTRY_CODE, --country_code COUNTRY_CODE
                        Country code(ISO 3166-1/ISO 3166-2) to avoid notifications on holidays.
```

#### Systemd units

The `systemd` directory provides units examples if you want to run this tool on a schedule.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Calendar Versioning](https://calver.org/).

## [Unreleased]

## [2022.3] - 2022-03-10
### Added
- SMTP configuration options to allow external server usage
- Support for SMTPS and SMTP authetication
- Systemd units examples
### Changed
- Code updated to Python 3.10
### Fixed
- Fixed selenium and workalendar deprecated code

## [2019.1] - 2019-05-15
### Added
- Use workalendar to avoid notifications on holidays
### Changed
- Better webdriver checking and error handling
### Fixed
- Fixed webdriver_path argument in README

## [2019.1b2] - 2019-01-15
### Fixed
- Update branch name for documentation links

## [2019.1b1] - 2019-01-15
### Changed
- Rewrite the project in Python 3
- Made as a Python package
- Docs using Sphinx
- Arguments can be provided on the command-line or via a configuration file
- Users can customize the screenshot size
- The script auto-detects if ChromeDriver's binary is present in PATH


traffic_info.sh
#!/bin/bash

# This program is free software: you can redistribute it and/or modify        
# it under the terms of the GNU General Public License as published by        
# the Free Software Foundation, either version 3 of the License, or           
# (at your option) any later version.                                         
#                                                                             
# This program is distributed in the hope that it will be useful,             
# but WITHOUT ANY WARRANTY; without even the implied warranty of              
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               
# GNU General Public License for more details.                                
#                                                                             
# You should have received a copy of the GNU General Public License           
# along with this program.  If not, see <http://www.gnu.org/licenses/>        

MAILTO=""
MAILFROM=""
LATITUDE=
LONGITUDE=
ZOOM=16

SCRIPT_DIR=$(dirname $0)
MAP_FILE="${SCRIPT_DIR}/map.html"
MAIL_CONTENT="${SCRIPT_DIR}/mail_content"
LOG_FILE="${SCRIPT_DIR}/webkit2png.log"

TEMP_DIR=$(mktemp -d)

#xvfb-run -a webkit2png --geometry=1280 720 --output=${TEMP_DIR}/map.png --wait=5 --feature=javascript --log=${LOG_FILE} file://${MAP_FILE}
sed -e "s|%%LAT%%|${LATITUDE}|" -e "s|%%LNG%%|${LONGITUDE}|" -e "s|%%ZOOM%%|${ZOOM}|" ${MAP_FILE} > ${TEMP_DIR}/map.html
xvfb-run -a webkit2png --geometry=1280 720 --output=${TEMP_DIR}/map.png --wait=5 --feature=javascript --log=${LOG_FILE} file://${TEMP_DIR}/map.html
#cp ${MAIL_CONTENT} ${TEMP_DIR}/mail
MAP_URL="https://www.google.fr/maps/@${LATITUDE},${LONGITUDE},${ZOOM}z/data=!5m1!1e1"
sed -e "s|%%MAP_URL%%|${MAP_URL}|" ${MAIL_CONTENT} > ${TEMP_DIR}/mail
uuencode -m ${TEMP_DIR}/map.png map.png >> ${TEMP_DIR}/mail
#cat ${TEMP_DIR}/mail | mailx -s "Traffic info" -a "Content-Type: multipart/related; boundary="boundary-example"; type="text/html"" -r ${MAILFROM} ${MAILTO}
cat ${TEMP_DIR}/mail | mailx -s "Traffic info" -a "Content-Type: multipart/related; boundary="traffic-info"; type="text/html"" -r ${MAILFROM} ${MAILTO}

rm -r ${TEMP_DIR}


traffic_info/core.py
"""Trafficinfo module."""
import logging
import os.path
import shutil
import smtplib
import tempfile
import time
from dataclasses import dataclass
from email.headerregistry import Address
from email.message import EmailMessage
from email.utils import make_msgid
from typing import Any, Dict

from selenium import webdriver
from selenium.webdriver.chrome.options import Options

from .utils import render_template

DIR = os.path.dirname(os.path.abspath(__file__))
CHROMEDRIVER_PATH = f"{DIR}/bin/chromedriver"

Context = Dict[str, Any]


@dataclass
class Location:
    """
    Location class.

    Args:
        latitude: Latitude of the center point of the map to make the screenshot.
        longitude: Longitude of the center point of the map.
        zoom: Google Maps zoom level, default 16.

    """

    latitude: float
    longitude: float
    zoom: int


class SMTPServer:
    """SMTPServer class."""

    def __init__(
        self,
        server: str = "localhost",
        port: int = 25,
        use_ssl: bool = False,
        login: str = None,
        password: str = None,
    ) -> None:
        """Initialize a SMTPServer object with the given options."""
        self.server = server
        self.port = port
        self.use_ssl = use_ssl
        self.login = login
        self.password = password
        self._smtp = None

    def __del__(self) -> None:
        """Cleanup stmp object."""
        if self._smtp is not None:
            self._smtp.quit()

    def connect(self) -> None:
        """Start SMTP connection."""
        if self.use_ssl:
            self._smtp = smtplib.SMTP_SSL(self.server, self.port)
        else:
            self._smtp = smtplib.SMTP(self.server, self.port)
        if self.login is not None and self.password is not None:
            self._smtp.login(self.login, self.password)

    def send_message(self, email: EmailMessage) -> None:
        """Send email message."""
        self._smtp.send_message(email)


class MapScreenshot:
    """
    MapScreenshot class.

    Args:
        webdriver_path: The path to the webdriver to use.
        api_key: Google Maps Javascript API key to take screenshots.
        width: Screenshot's width, default 1280.
        height: Screenshot's height, default 720.
        output_dir: The path to save the screenshot,
        if not specified a temporary directory will be created.

    """

    _template_dir: str = os.path.join(DIR, "templates")

    def __init__(
        self,
        webdriver_path: str,
        api_key: str = None,
        width: int = 1280,
        height: int = 720,
        output_dir: str = None,
    ) -> None:
        """Initialize a MapScreenshot object with the given options."""
        self._tmpdir: str = tempfile.mkdtemp()
        self.webdriver_path: str = webdriver_path
        self.api_key: str = api_key
        self.width: int = width
        self.height: int = height
        self.output_dir = output_dir if output_dir is not None else self._tmpdir
        self.path: str = None

    def take(self, location: Location) -> str:
        """
        Take a screenshot using chrome webdriver.

        Args:
            location: The location of the map's center point.

        Returns:
            The screenshot's full path.

        """
        context: Context = {
            "key": self.api_key,
            "latitude": location.latitude,
            "longitude": location.longitude,
            "zoom": location.zoom,
        }
        map_html = os.path.join(self._tmpdir, "map.html")
        render_template(os.path.join(self._template_dir, "map.j2"), context, map_html)

        options = Options()
        options.headless = True
        driver = webdriver.Chrome(
            executable_path=self.webdriver_path, chrome_options=options
        )
        driver.set_window_size(self.width, self.height)
        driver.get(f"file://{map_html}")
        time.sleep(5)
        self.path = os.path.join(self.output_dir, "map.png")
        driver.save_screenshot(self.path)
        driver.quit()
        return self.path

    def __del__(self) -> None:
        """Cleanup temporary files."""
        if os.path.isdir(self._tmpdir):
            shutil.rmtree(self._tmpdir)


def send_email(
    email_from: str,
    email_to: str,
    location: Location,
    screenshot: MapScreenshot,
    smtp_server: SMTPServer,
    template: str = None,
) -> None:
    """
    Send the traffic info email.

    Args:
        email_from: Email sender's address.
        email_to: Email recipient's address.
        location: The map's location.
        screenshot: The map's screenshot.
        template: The path to the email's Jinja2 template,
        templates/email.j2 if not specified.

    """
    if template is None:
        template = os.path.join(DIR, "templates", "email.j2")
    logger = logging.getLogger(__name__)
    map_cid = make_msgid()
    context: Context = {
        "url": f"https://www.google.fr/maps/@{location.latitude},"
        f"{location.longitude},{location.zoom}z/data=!5m1!1e1",
        "width": screenshot.width,
        "height": screenshot.height,
        "map_cid": map_cid[1:-1],
    }
    content = f"""
    Today's traffic conditions.
    {context["url"]}
    Have a safe trip back home!
    """
    html = render_template(template, context)
    email = EmailMessage()
    email["Subject"] = "Traffic info"
    email["From"] = Address("Traffic info", addr_spec=email_from)
    email["To"] = email_to
    email.set_content(content)
    email.add_alternative(html, subtype="html")
    with open(screenshot.path, "rb") as img:
        email.get_payload()[1].add_related(img.read(), "image", "png", cid=map_cid)

    try:
        smtp_server.connect()
    except (ConnectionRefusedError, smtplib.SMTPAuthenticationError) as exception:
        logger.error("Unable to connect to SMTP server(%s)", exception)

    try:
        smtp_server.send_message(email)
    except smtplib.SMTPException as exception:
        logger.error("Unable to send email(%s)", exception)


traffic_info/utils.py
"""Utils for traffic_info package."""
import shutil
from typing import Any, Dict

import jinja2

Context = Dict[str, Any]


def get_chromedriver_path() -> str:
    """
    Get ChromeDriver's binary path.

    Returns:
        The ChromeDriver's binary path, None otherwise.

    """
    chromedriver_path = shutil.which("chromedriver")
    return chromedriver_path


def render_template(template: str, context: Context, output: str = None) -> str:
    """
    Render a Jinja2 template.

    Args:
        template: The path to the Jinja2 template.
        context: The Jinja2 context.
        output: The path to save the rendered file.

    Returns:
        The rendered template if output is None or the path to the rendered file.

    """
    with open(template, mode="r", encoding="utf-8") as template_file:
        jinja_template = jinja2.Template(template_file.read())

    render = jinja_template.render(context)
    if output is None:
        return render
    with open(output, mode="w", encoding="utf-8") as render_file:
        render_file.write(render)
    return output


traffic_info/exceptions.py
"""Traffic Info's exceptions."""


class TrafficInfoError(Exception):
    """Base class for exceptions in this module."""


class WebdriverNotFoundError(TrafficInfoError):
    """
    Exception raised when the webdriver is not found.

    Attributes:
        path: The webdriver's path.

    """

    def __init__(self, path):
        """Class init."""
        super().__init__()
        msg = "Webdriver not found"
        if path is not None:
            msg += f" ({path})"
        self.msg = f"{msg}."


class NotExecutableError(TrafficInfoError):
    """
    Raised when the webdriver file is not executable.

    Attributes:
        path: The webdriver's path.

    """

    def __init__(self, path):
        """Class init."""
        super().__init__()
        self.msg = f"The file {path} is not executable."


traffic_info/__init__.py
"""traffic_info module."""
import datetime
import logging
import os
import sys

import configargparse

from selenium.common.exceptions import WebDriverException

from workalendar.registry import registry

from .__version__ import __version__
from .core import Location, MapScreenshot, SMTPServer, send_email
from .exceptions import NotExecutableError, WebdriverNotFoundError
from .utils import get_chromedriver_path


def check_webdriver_path(webdriver: str = None) -> str:
    """
    Check if the webdriver exist and is executable.

    Args:
        webdriver: The webdriver's path.

    Returns:
        The verified path.

    """
    if webdriver is None:
        webdriver = get_chromedriver_path()
        if webdriver is None:
            raise WebdriverNotFoundError(webdriver)
    if not os.path.isfile(webdriver):
        raise WebdriverNotFoundError(webdriver)
    if not os.access(webdriver, os.X_OK):
        raise NotExecutableError(webdriver)
    return webdriver


def parse_args() -> configargparse.Namespace:
    """Parse command-line arguments."""
    parser = configargparse.ArgParser()
    parser.add_argument(
        "-c", "--config-file", is_config_file=True, help="Config file path."
    )
    parser.add_argument("-d", "--webdriver_path", help="ChromeDriver's path.")
    parser.add_argument("-k", "--api_key", help="Google Maps Javascript API key.")
    parser.add_argument(
        "-l",
        "--latitude",
        required=True,
        type=float,
        help="Latitude of the center point of the map.",
    )
    parser.add_argument(
        "-L",
        "--longitude",
        required=True,
        type=float,
        help="Longitude of the center point of the map.",
    )
    parser.add_argument("-z", "--zoom", type=int, help="Google Maps zoom level.")
    parser.add_argument(
        "-f", "--email_from", required=True, help="Email sender’s address."
    )
    parser.add_argument(
        "-t", "--email_to", required=True, help="Email recipient’s address."
    )
    parser.add_argument(
        "-s",
        "--smtp_server",
        type=str,
        default="localhost",
        help="SMTP server’s address.",
    )
    parser.add_argument(
        "-p", "--smtp_port", type=int, default=25, help="SMTP server’s port."
    )
    parser.add_argument("-S", "--smtp_use_ssl", action="store_true", help="Use SMTPS.")
    parser.add_argument("-u", "--smtp_login", type=str, help="SMTP server’s login.")
    parser.add_argument(
        "-w", "--smtp_password", type=str, help="SMTP server’s password."
    )
    parser.add_argument(
        "-W", "--screenshot_width", type=int, help="Screenshot’s width."
    )
    parser.add_argument(
        "-H", "--screenshot_height", type=int, help="Screenshot’s height."
    )
    parser.add_argument(
        "-C",
        "--country_code",
        type=str,
        help="Country code(ISO 3166-1/ISO 3166-2) to avoid notifications on holidays.",
    )

    return parser.parse_args()


def run() -> None:
    """Run traffic info from command line."""
    logger = logging.getLogger(__name__)
    logger.setLevel(logging.INFO)

    options = parse_args()
    if options.country_code:
        calendar_class = registry.get(options.country_code.upper())
        if not calendar_class:
            logger.error("Invalid country code")
            sys.exit(1)
        calendar = calendar_class()
        if calendar.is_holiday(datetime.datetime.now()):
            # Enjoy your holiday! :)
            sys.exit()

    location_keys = ["latitude", "longitude", "zoom"]
    screenshot_keys = ["api_key", "webdriver_path", "width", "height"]

    location_params = {
        k: v for k, v in vars(options).items() if v is not None and k in location_keys
    }
    screenshot_params = {
        k: v for k, v in vars(options).items() if v is not None and k in screenshot_keys
    }
    try:
        screenshot_params["webdriver_path"] = check_webdriver_path(
            options.webdriver_path
        )
    except (NotExecutableError, WebdriverNotFoundError) as exc:
        logger.error(exc.msg)
        sys.exit(1)

    location = Location(**location_params)
    screenshot = MapScreenshot(**screenshot_params)
    try:
        screenshot.take(location)
    except WebDriverException as exc:
        logger.error(exc.msg)
        sys.exit(1)
    smtp_server = SMTPServer(
        options.smtp_server,
        options.smtp_port,
        options.smtp_use_ssl,
        options.smtp_login,
        options.smtp_password,
    )
    send_email(
        options.email_from,
        options.email_to,
        location,
        screenshot,
        smtp_server,
    )


__all__ = ["__version__", "Location", "MapScreenshot", "run", "send_email"]
